test/cpp/api/inference_mode.cpp

```diff
@@ -304,33 +304,54 @@ TEST(InferenceModeTest, TestMixInferenceAndNormalTensorFunctionalOp) {
     torch::Tensor c;
     {
       InferenceMode guard;
-      c = torch::ones({1, 2, 3});
+      c = torch::ones({1, 2, 3}).set_requires_grad(requires_grad);
     }
 
-    ASSERT_THROWS_WITH(c.add(s), // go through kernels: VariableType(ERROR!), InplaceOrView(fallthrough), CPU
-      "Inference tensor cannot participate in autograd")
+    // add(Tensor, Tensor) is safe with inference tensor since it doesn't save any variable for backward.
+    torch::Tensor out = c.add(s);  // go through kernels: VariableType, InplaceOrView(fallthrough), CPU
+    ASSERT_FALSE(is_inference_tensor(out));
+    ASSERT_EQ(out.requires_grad(), requires_grad);
+    if (requires_grad) {
+      // leaf inference tensor with requires_grad=true can still have gradient.
+      out.backward(torch::ones_like(out));
+      assert_tensor_equal(c.grad(), torch::ones_like(c));
```

Comment:
ngl this is pretty weird (but maybe ok). One thing to note is that this means that we don't have an invariant that inference tensors are guaranteed not to have AutogradMeta (since the gradient is stored there)

Reply:
Yea I feel inference tensor eventually can participate in autograd properly as long none of its version/view invariants is violated. Wanted to leave this test here since it's naturally what people would expect from inference tensor participating autograd I think. :)